File: Daily Challenge/February 2025/Letter Tile Possibilities.cpp
class Solution {
public:
  int numTilePossibilities(string tiles) {
    unordered_set<string> sequences;
    vector<bool> used(tiles.size(), false);

    string current = "";
    generateSequences(tiles, used, sequences, current);

    return sequences.size() - 1;
  }

private:
  void generateSequences(string& tiles, vector<bool>& used, unordered_set<string>& sequences, string& current) {
    
    sequences.insert(current);

    for (int i = 0; i < tiles.size(); ++i) {
      if (used[i]) {
        continue;
      }

      used[i] = true;
      current.push_back(tiles[i]);
      generateSequences(tiles, used, sequences, current);
      current.pop_back();
      used[i] = false;
    }
  }
};

File: Daily Challenge/February 2025/Letter Tile Possibilities.cs
public class Solution {
  public int NumTilePossibilities(string tiles) {
    HashSet<string> sequences = new HashSet<string>();
    bool[] used = new bool[tiles.Length];

    GenerateSequnces(tiles, used, "", sequences);

    return sequences.Count - 1;
  }

  private void GenerateSequnces(string tiles, bool[] used, string current, HashSet<string> sequences) {
    sequences.Add(current);

    for (int i = 0; i < tiles.Length; i++) {
      if (used[i]) continue;

      used[i] = true;
      GenerateSequnces(tiles, used, current + tiles[i], sequences);
      used[i] = false;
    }
  }
}